Uniquely identify overloaded routines

fourls opened the pull request:
Currently, routines are identified only by their name, which leads to a number of issues including the inability to link to specific overloads, and duplicate ID issues with HTML output. #146 introduced another problem stemming from this: if you have overloaded methods inheriting a description from a base class, the same description is displayed for all of them. 

This PR implements overload support, including

* Unique routine anchors by parameter (e.g. `procedure MyProc(Message: string)` has anchor `#MyProc(string)`)
* Correct inherited description resolution for overloaded methods
* Optional specification of routine parameters in links (e.g. `@link(MyProc(string) A link to MyProc)`)

An example of the new syntax is below:

```pascal
{
@link(MyProc any overload)
@link(MyProc(Integer,string) the int/string overload)
@link(MyProc(string) the string overload) 
}
type TMyClass = class(TObject)
  procedure MyProc(MyInt: Integer; MyStr: string); overload;
  procedure MyProc(MyStr: string); overload;
end;
```

I see there has been discussion over the best way to do this in [ToDoOverloadingLinking](https://github.com/pasdoc/pasdoc/wiki/ToDoOverloadingLinking). The way I've gone about it is by adding a new `Signature` property to `TBaseItem`, which represents the unique identifier for that item. This just returns `Name` in all places except `TPasRoutine`, in which it returns `<name>(<param 1 type>,<param 2 type>,<etc.>)`. The `Signature` property is used for generating anchors and links.

`TPasRoutines` now also contains a second hashmap for mapping simple names to the signature of the most recently parsed item with that name. If `FindListItem` is passed a string that is not in the main hashmap, it will check the second hashmap to see if it's in there, and resolve it that way.

Caveats: 

* The `@link` first word cannot have spaces, which means `@link(MyProc(Integer, string)` does NOT work. If this is important I can have a look at fixing it now, otherwise I'll have a look at it in the future. **FIXED NOW**
* There may be some areas I have missed replacing `Name` with `Signature` (although I have manually tested HTML output pretty extensively and it seems correct!) 
* I also don't have any LaTeX tooling so I've only been able to implement this for HTML - changing other output formats should be very simple, though.

Fr0sT-Brutal commented:
Wow that's really helpful!
Regarding @_link and spaces: we could scan for brackets and add everything between them as link's part. Then remove spaces and thus generate proper anchor.

fourls commented:
Very true - I've done that now and seems to work great!

michaliskambi commented on source/component/PasDoc_GenHtml.pas:
```diff
@@ -1060,7 +1060,12 @@ procedure TGenericHTMLDocGenerator.WriteItemTableRow(
   WriteStartOfTableCell('itemcode');
 
-  if MakeAnchor then WriteAnchor(Item.Name);
+  if MakeAnchor then begin
```
From what I see, `TBaseItem.Signature` returns just `Name`. 

So this code should not need any special condition `    if Item is TPasRoutine then`, and can just do

```
if MakeAnchor then WriteAnchor(Item.Signature);
```

michaliskambi commented on source/component/PasDoc_Items.pas:
```diff
@@ -2701,6 +2803,7 @@ procedure TPasUnits.SetUnitAt(const AIndex: Integer; const Value: TPasUnit);
 begin
   inherited;
+  FParamTypes := TStringVector.Create;
 end;
```
From a quick glance, this may be a memory leak -- you never do `FreeAndNil(FParamTypes)` .

michaliskambi reviewed (changes requested):
Great feature! We missed it for a long time :) The ability to use `@link` to link to a specific overload was in particular requested a few times.

It was also preventing HTML validity, as we had duplicate ids in HTML.

I found 2 possible small issues, made comments at code lines below.

3rd note: I see that right now this will result in HTML ids with `(` and `)` characters. While HTML 5 allows it ( https://html.spec.whatwg.org/multipage/dom.html#the-id-attribute . found from https://stackoverflow.com/questions/70579/what-are-valid-values-for-the-id-attribute-in-html ) :

> There are no other restrictions on what form an ID can take; in particular, IDs can consist of just digits, start with a digit, start with an underscore, consist of just punctuation, etc.

... but the recommendation from MDN ( https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id ) is to limit yourself:

> Technically, the value for an id attribute may contain any character, except [whitespace](https://developer.mozilla.org/en-US/docs/Glossary/Whitespace) characters. However, to avoid inadvertent errors, only [ASCII](https://developer.mozilla.org/en-US/docs/Glossary/ASCII) letters, digits, '_', and '-' should be used and the value for an id attribute should start with a letter. For example, . has a special meaning in CSS (it acts as a [class selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Class_selectors)). Unless you are careful to escape it in the CSS, it won't be recognized as part of the value of an id attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.

I'd say to add a method like `SignatureToId` used in HTML generator that converts `TBaseItem.Signature` to HTML id, and changes anything unsafe (like `(`, `)`) to underscore.

4th note: check how does the HTML validation change after this :)

I did play with enabling HTML validation automatically also from GH Actions, you can see my comments in `.github/workflows/build.yml`  https://github.com/pasdoc/pasdoc/blob/master/.github/workflows/build.yml#L48 . 

You can uncomment the lines relevant to `HTML5Validator` and see how does the HTML validation change -- the ideal result is that 3x errors `error: Duplicate ID` disappear, and no new issue appears. 

You can also install the validator listed there manually, either as https://github.com/validator/validator or https://github.com/svenkreiss/html5validator , and run it locally (without relying on GH Actions).

fourls commented:
Good points, thanks Michalis!

> I found 2 possible small issues, made comments at code lines below.

Fixed :)

> I'd say to add a method like `SignatureToId` used in HTML generator that converts `TBaseItem.Signature` to HTML id, and changes anything unsafe (like `(`, `)`) to underscore.

Done - I've converted the unsafe characters to hyphens instead, as identifiers can contain underscores. 

I've noticed there are a couple of cases in which the method names can contain unsafe characters, such as operator overloads that can have names like `*`, `:=`, `/`, etc. I don't think it's worth bothering with for now, but something to keep in mind.

> 4th note: check how does the HTML validation change after this :)

Two of the three duplicate ID errors disappear - the third one is from two enums that share a value name, so it makes sense that it would not be affected by this change. 

My v.Nu output was different to what you documented in the `build.yml`, but I suspect that v.Nu is misreading the text encoding of the files and it's not actually a problem - I checked and it was similar on master (see the diff of my v.Nu output between master and this PR branch):

![image](https://user-images.githubusercontent.com/28214458/205524929-6e95d865-3f51-401b-a84d-184235c27ea2.png)

michaliskambi commented:
Thank you, everything looks good!

> My v.Nu output was different to what you documented in the build.yml, but I suspect that v.Nu is misreading the text encoding of the files and it's not actually a problem - I checked and it was similar on master (see the diff of my v.Nu output between master and this PR branch):

Hm, weird, for me I can get local output that matches the GH Actions output using just

```
html5validator --root testcases_output/html/
```

after installing it following https://github.com/svenkreiss/html5validator (I used `pip install html5validator` on Linux). Oh well, doesn't matter, since the diff is correct -- great, 2 issues less in validation :) Thanks for checking!
